### ISSUE TITLE
Deleting an organization is very slow

### ISSUE THREAD

**Issue by Zhao-Andy** [forem/forem#12575]
**Describe the bug**
When you try to delete an organization, via any method `Organization.find(id).destroy` or through the organization settings, it takes a really long time -- at least on dev.to.

**To Reproduce**
1. Try to delete an organization (on dev.to)
2. Find that it doesn't work a lot and you're stuck with belonging to the organization you're trying to delete

**Expected behavior**
A user can delete their organization.

**Additional context**
So I believe that the underlying query that is halting this process is:
https://github.com/forem/forem/blob/9ef1534d83b7cf2e62003eb226cf6912f0c67d3d/app/models/organization.rb#L30

There could be other queries that are slowing it down, but I don't think we have any tables that are as large as `notifications`.

For some reason, `Notification.where(organization_id: id)` will takes a _really_ long time to run. I'm not sure why the query is so slow. 

Here are some queries I ran that might help:
```
# EXPLAIN SELECT * FROM notifications WHERE organization_id = 1;

                                      QUERY PLAN
------------------------------------------------------------------------------------------
 Index Scan using index_notifications_user_id_organization_id_notifiable_action on notifications 
   (cost=0.14..3167530.39 rows=150 width=1056)
   Index Cond: (organization_id = 1)
(2 rows)
```

```
# EXPLAIN ANALYZE SELECT * FROM notifications where organization_id = 1;

                                      QUERY PLAN
------------------------------------------------------------------------------------------
 Index Scan using index_notifications_user_id_organization_id_notifiable_action on notifications 
   (cost=0.14..3167530.39 rows=150 width=1056) (actual time=10679.013..10705.613 rows=2369 loops=1)
   Index Cond: (organization_id = 1)
 Planning Time: 0.182 ms
 Execution Time: 10705.765 ms
(4 rows)
```

The fix could be as simple as changing it from `dependent: :destroy` to `dependent: :delete_all`, or adding an index for `notifications.organization_id`.

**Comment by Zhao-Andy** [comment]
cc SQL experts @citizen428 @rhymes 😄

**Comment by rhymes** [comment]
Hi @Zhao-Andy, I think the main reason here is that `notifications` is a huge table. Deleting an organization is not that different from deleting a user in those regards and we moved away from `user.destroy` to `Users::Delete.call(user)` to be able to control those steps.

According to the code `organizations` have a `destroy` dependency on `display_ads`, `listings`, `notifications`, `profile_pins` and more. They also have `nullify` dependencies on `articles` which is a sizable table to enumerate so that `update articles set organization_id = NULL where organization_id = 123` can run successfully.

I agree with you that the major culprit might be the `notifications` table. I saw you submit a PR to switch to `.delete_all`. I think that's a sensible solutions as there are no `after_destroy` callbacks  on the `Notification` table. 

BTW the index for `organization_id` already exists, as you can see in your query plan, PostgreSQL is recycling `index_notifications_user_id_organization_id_notifiable_action` which contains all the organization ids already.

It's just too big of a table to run N separate `DELETE` statements (what `.destroy_all` does) vs running one (what `.delete_all` does)

**Comment by Zhao-Andy** [comment]
Unfortunately changing it from `.delete_all` did not reduce the the time enough and the process still times out :( I wonder if the index we currently have `index_notifications_user_id_organization_id_notifiable_action` is not efficient enough here, and if we should add a new index specifically for deleting organizations `index_notifications_organization_id`

```ruby
class AddOrganizationIdIndexToNotifications
  def change
    add_index :notifications, :organization_id # or whatever the syntax
  end
end
```

But perhaps the real issue is that the notifications table has gotten too large...

**Comment by rhymes** [comment]
@Zhao-Andy mmm we can try but I don't think adding another index will solve the issue (the existing one already contains all the data). Another option is to call `organization.notifications.in_batches.delete_all` before the `organization.destroy`

I fear the long term solution here is to have an `Organizations::DeleteWorker` that calls `Organizations::Delete.call(organization)` and in turn deletes `organizations` asynchronously by also removing notifications in batches

**Comment by citizen428** [comment]
@Zhao-Andy According to this [DBA StackExchange answer](https://dba.stackexchange.com/a/138270/141) the `ON DELETE CASCADE` triggers in combination with missing indexes could indeed be the culprit.

That said, I agree with @rhymes, we probably want to model org deletion after user deletion and have it handled by a service.

**Comment by AlbertoPdRF** [comment]
Maybe the new (Rails 6.1) [`dependent: :destroy_async`](https://github.com/rails/rails/pull/40157) could be useful here?

**Comment by citizen428** [comment]
@AlbertoPdRF Yes, it most likely would be, alas we're not quite there with the update yet, see the tracking PR: https://github.com/forem/forem/pull/11333

**Comment by mstruve** [comment]
I think modeling org deletion off of user deletion and creating a service for it like @citizen428 mentions is the way to go for this. It also means removing more callbacks 💃

**Comment by Zhao-Andy** [comment]
Resolved by #13388, thanks @lightalloy!
